internal/core/database_api/athena/driver/api/athena.go

```diff
@@ -0,0 +1,206 @@
+	output := &models.GetQueryStatusOutput{}
+
+	var err error
+	defer func() {
+		if err != nil {
+			output.ErrorMessage = "GetQueryStatus failed" // simple error message for lambda caller
+		}
+	}()
+
+	executionStatus, err := awsathena.Status(athenaClient, input.QueryID)
+	if err != nil {
+		return output, err
+	}
+	output.Status = getQueryStatus(executionStatus)
+	return output, nil
+}
+
+func (API) GetQueryResults(input *models.GetQueryResultsInput) (*models.GetQueryResultsOutput, error) {
+	output := &models.GetQueryResultsOutput{}
+
+	var err error
+	defer func() {
+		if err != nil {
+			output.ErrorMessage = "GetQueryResults failed" // simple error message for lambda caller
+		}
+	}()
+
+	executionStatus, err := awsathena.Status(athenaClient, input.QueryID)
+	if err != nil {
+		return output, err
+	}
+	output.Status = getQueryStatus(executionStatus)
+
+	if output.Status == models.QuerySucceeded {
+		var nextToken *string
+		if input.PaginationToken != nil { // paging thru results
+			nextToken = input.PaginationToken
+		}
+		err = getQueryResults(athenaClient, executionStatus, output, nextToken, input.ResultsMaxPageSize)
+		if err != nil {
+			return output, err
+		}
+	}
+
+	return output, nil
+}
+
+func getQueryStatus(executionStatus *athena.GetQueryExecutionOutput) string {
+	switch *executionStatus.QueryExecution.Status.State {
+	case
+		athena.QueryExecutionStateSucceeded:
+		return models.QuerySucceeded
+	case
+		// failure modes
+		athena.QueryExecutionStateFailed,
+		athena.QueryExecutionStateCancelled:
+		return models.QueryFailed
+	case
+		// still going
+		athena.QueryExecutionStateRunning,
+		athena.QueryExecutionStateQueued:
+		return models.QueryRunning
+	default:
+		panic("unknown athena status: " + *executionStatus.QueryExecution.Status.State)
+	}
+}
+
+func getQueryResults(client athenaiface.AthenaAPI, executionStatus *athena.GetQueryExecutionOutput,
+	output *models.GetQueryResultsOutput, nextToken *string, maxResults *int64) (err error) {
+
+	queryResult, err := awsathena.Results(client, executionStatus, nextToken, maxResults)
+	if err != nil {
+		return err
+	}
+	err = collectResults(queryResult, output, maxResults)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func collectResults(queryResult *athena.GetQueryResultsOutput, output *models.GetQueryResultsOutput, maxResults *int64) (err error) {
+	for _, row := range queryResult.ResultSet.Rows {
+		var columns []*models.Column
+		for _, col := range row.Data {
+			columns = append(columns, &models.Column{
+				Value: *col.VarCharValue,
+			})
+		}
+		output.Rows = append(output.Rows, &models.Row{Columns: columns})
+	}
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	output.NumRows = len(queryResult.ResultSet.Rows)
+	if output.NumRows > 0 && (maxResults == nil || int(*maxResults) == output.NumRows) { // could be more!
+		output.PaginationToken = queryResult.NextToken
+	} else {
+		output.PaginationToken = nil // no more
+	}
```

Comment:
Why not just

```suggestion
output.PaginationToken = queryResult.NextToken
```

And let the SDK tell you when there might be another page?

Reply:
Because it always has one ... I'll double check that, but during testing I found that annoying. Maybe I'm wrong.

Reply:
Tested again and the simple form works. My tests must have been borked at the time. Fixed